soc/intel/intel_adsp/tools/acetool.py

```diff
@@ -538,15 +551,77 @@ def winstream_read(last_seq):
             return (seq, "")
         copy = (end - behind) % wlen
         suffix = min(behind, wlen - copy)
-        result = win_read(16 + copy, suffix)
+        result = win_read(base, 16 + copy, suffix)
         if suffix < behind:
-            result += win_read(16, behind - suffix)
-        (wlen, start1, end, seq1) = win_hdr()
+            result += win_read(base, 16, behind - suffix)
+        (wlen, start1, end, seq1) = win_hdr(base)
         if start1 == start and seq1 == seq:
             # Best effort attempt at decoding, replacing unusable characters
             # Found to be useful when it really goes wrong
             return (seq, result.decode("utf-8", "replace"))
 
+def idx_mod(wlen, idx):
+    if idx >= wlen:
+        return idx - wlen
+    return idx
+
+def idx_sub(wlen, a, b):
+    return idx_mod(wlen, a + (wlen - b))
+
+# Python implementation of the same algorithm in sys_winstream_write(),
+# see there for details.
+def winstream_write(base, msg):
+    (wlen, start, end, seq) = win_hdr(base)
+    if wlen > SHELL_MAX_VALID_SLOT_SIZE:
+        log.debug("DSP powered off at winstream_write")
+        return
+    if wlen == 0:
+        return
+    lenmsg = len(msg)
+    lenmsg0 = lenmsg
+    if len(msg) > wlen + 1:
+        start = end
+        lenmsg = wlen - 1
+    lenmsg = min(lenmsg, wlen)
+    if seq != 0:
+        avail = (wlen - 1) - idx_sub(wlen, end, start)
+        if lenmsg > avail:
+            start = idx_mod(wlen, start + (lenmsg - avail))
+    if lenmsg < lenmsg0:
+        start = end
+        drop = lenmsg0 - lenmsg
+        msg = msg[drop : lenmsg - drop]
+    suffix = min(lenmsg, wlen - end)
+    for c in range(0, suffix):
+        bar4_mmap[base + 16 + end + c] = ord(msg[c])
+    if lenmsg > suffix:
+        for c in range(0, lenmsg - suffix):
+            bar4_mmap[base + 16 + c] = ord(msg[suffix + c])
+    end = idx_mod(wlen, end + lenmsg)
+    seq += lenmsg0
+    update_hdr = struct.pack("<IIII", wlen, start, end, seq)
+    # write via array does not seem to work, so writing bytes one by one
```

Comment:
Do note that this is actually a historical reliability problem with the python implementation.  The algorithm is lockless and relies on the values changing atomically[1].  If we're doing bytewise I/O the firmware can end up seeing corrupt values, which will turn up as runtime errors (ideally a stream reset unless winstream itself has bugs).

For the specific case of a python script doing debug or logging access... it's probably OK.  (Though now that we're writing to the firmware there are new opportunities for hijinx).   If we want to do this right we probably need to be doing ctypes access to the shared memory buffer and not the bytearray stuff.

[1] And in the right order.  In the case of write() (because winstream doesn't promise stream reliability in the case of overrun) all that's required is that the "end" field be updated only after all the bytes have been copied in and the start/len fields updated.

Reply:
Changed this to 32bit writes via mmap slice in V4. Not sure if this is guaranteed to work, but a 32bit mmap write should be fairly reliable. I did not remember to put "end" to last, so I'll take a look at that still.